src/translator/quality_estimator.h

```diff
@@ -0,0 +1,222 @@
+  /// \f$w\f$ is the model vector with learnt weights
+  ///
+  /// 3) We apply the sigmoid function to the result
+  ///
+  /// Notice, however, that for the first two steps we can do the following:
+  ///
+  /// \f{align*}{
+  /// \sum_{i=1}^{D}{ w_i z_i} &= \mathbf{w^T}\left(\mathbf{\sigma^{-1}} \odot (\mathbf{x} - \mathbf{\mu})\right) \text{
+  /// //
+  /// we are just vectoring step 1}\\
+  ///      &= \sum_{i=1}^{D}{\sigma_i^{-1} w_i (x_i - \mu_i)} \\
+  ///      &= \sum_{i=1}^{D}{\sigma_i^{-1} w_ix_i - \sigma_i^{-1} w_i \mu_i} \\
+  ///      &= \sum_{i=1}^{D}{\left(\sigma_i^{-1} w_i\right)x_i - \left(\sigma_i^{-1} w_i \mu_i\right)}
+  /// \f}
+  /// Then, \f$(\sigma_i^{-1} w_i \mu_i)\f$ can be precomputed without any dependence on inference data. This is done by
+  /// the variable \f$\textit{constantFactor_}\f$ and \f$\textit{intercept_}\f$ in the code.
+  ///
+  /// @param [in] features: A Matrix struct of features. For a defintion what features currently means, please refer to
+  /// `extractFeatures` method in `quality_estimator.cpp`
+  std::vector<float> predict(const Matrix &features) const;
+
+ private:
+  Scale scale_;
+  Array coefficients_;
+  float intercept_;
+  Array coefficientsByStds_;
+  float constantFactor_ = 0.0;
+
+  // Number of parameters with dimension - Scale(stds, means) and coefficients
+  static constexpr const size_t numLrParamsWithDimension_ = 3;
+  // Number of intercept values
+  static constexpr const size_t numIntercept_ = 1;
+
+  /// construct the struct SentenceQualityEstimate
+  /// @param [in] logProbs: the log probabilities given by an translation model
+  /// @param [in] target: AnnotatedText target value
+  /// @param [in] sentenceIdx: the id of a candidate sentence
+  Response::SentenceQualityScore computeSentenceScores(const std::vector<float> &logProbs, const AnnotatedText &target,
+                                                       const size_t sentenceIdx) const;
+
+  Matrix extractFeatures(const std::vector<SubwordRange> &wordIndices, const std::vector<float> &logProbs) const;
+};
+
+/// createQualityEstimator model takes an `AlignedMemory`, which is the return from `getQualityEstimatorModel`.
+///
+/// `getQualityEstimatorModel` contains two different implementations, one when the `quality` argument has some value as
+/// a possible `Options` and where it does not.
+///
+/// If a non `quality` option is provided, then by default, it uses the UnsupervisedQualityEstimator implementation.
+///
+/// If a value is passed to the `quality` argument, the model file is read and converted into an `AlignedMemory`
+/// structure, which instantiates a QualityEstimator object.
+
+/// @param [in] qualityFileMemory: An `AlignedMemory` which is created by parsing a QE model binary file through
+/// getQualityEstimatorModel
+inline std::shared_ptr<QualityEstimator> createQualityEstimator(const AlignedMemory &qualityFileMemory) {
```

Comment:
Based on this signature the QE model type has to be inferred from contents of the file. There's one more type coming in? How are you intending to handle switching construction?